Homework 2

asp2403 opened the pull request:
Работа над ошибками

asp2403 committed:
1. Убрана лишняя инициализация бинов

2. Добавлены новые исключения с сообщениями
3.  Новые методы в IOServiceImpl (форматирование и чтение char)
4. Проверки перенесены в IOServiceImpl, в случае неправильных входных данных бросаются исключения
5. Рефакторинг AppRunner

asp2403 committed:
Вынесение части функционала TestingService в QuestionService и ScoreService. Рефакторинг AppRunner

asp2403 committed:
User убран из сервисов и перенесен в AppRunner. В IOService добавлены методы для корректного чтения входных параметров, все проверки вынесены из AppRunner в IOServuce. Лишние исключения удалены.

stvort commented on hw2/src/main/java/ru/otus/homework/popov/service/AppRunner.java:
```diff
@@ -5,30 +5,45 @@
     private final TestingService testingService;
     private final IOService ioService;
     private final QuestionConverter questionConverter;
+    private User user;
```
>Лучше обойтись без состояния, если это возможно. Тут возможно)

Тут тоже)

stvort commented on hw2/src/main/java/ru/otus/homework/popov/service/AppRunner.java:
```diff
@@ -41,33 +56,41 @@ public AppRunner(
-            sayGoodBy();
-            return;
+        sayHello();
+        if (registerUser()) {
+            run();
         }
```
Например так: `registerUser().ifPresent(this::run)`, где первый метод возвращает `Optional<User>`

stvort commented on hw2/src/main/java/ru/otus/homework/popov/service/AppRunner.java:
```diff
@@ -77,45 +100,62 @@ private void sayGoodBy() {
+    }
+
+
+    private boolean runTest() {
+        testingService.startTest();
+        if (!answerQuestions()) {
+            return false;
+        }
+        printScore();
+        return true;
+    }
```
Это все тоже подразумевает состояние. Обратите внимание на код в последнем комментарии (дополнил одной строкой):
```
User user = userService.registerUser();
TestingResult res = testingService.testUser(user)
testResultPrinter.print(res);
sayGoodBye(res.getUser());
```
В этом коде есть результат тестирования и юзер. И они оба являются локальными переменными. Вот так может выглядеть метод `testUser`
```
TestingResult tesingResult = new TestingResult(user);
List<Question> questions = questionsService.getQuestion();
for (Question q: questions) {
    boolean answerResult = askQuestion(q);
    tesingResult.applyAnswer(answerResult);
}
return tesingResult;
```
Тут вопросы и результат тоже локальные переменные. Состояние нигде не хранится в полях. Только локальные переменные, возвращаемые значения и аргументы методов

asp2403 replied:
Исправил. А чем так плохи состояния? Теперь локальные переменные приходится протаскивать через параметры вызываемых методов.

stvort commented on hw2/src/main/java/ru/otus/homework/popov/service/AppRunner.java:
```diff
@@ -41,33 +56,41 @@ public AppRunner(
+        sayGoodBy();
+    }
+
+    private void run() {
+        do {
+            if (!runTest()) {
+                return;
+            }
+            if (!wantTryAgain()) {
+                return;
+            }
+        } while (true);
     }
```
Пока код достаточно сложный для восприятия. 
Что значит `if (!runTest())`? Если не получилось пройти тест? 
Внутри это `!answerQuestions()`, т.е. не вышло ответить на вопросы? 
Проваливаемся далее `!processUserAnswer(question)`, т.е. на вопросы не получилось ответить т.к. не получилось обработать ответ? 
И только уже внутри оказывается, что не получилось обработать ответ это когда пользователь решил выйти

asp2403 replied:
Согласен, исправил

asp2403 committed:
1. Убраны состояния из AppRunner

2. Добавлен новый доменный класс TestingResult
3. Для лучшено восприятия кода вместо возвращаемых значений в методах обработки тестирования используется флаг isTerminated
4. Выделены новые сервисы TestingResultPrinter, AppConfig. Удален сервис ScoreService
5. Сообщения вынесены в отдельный класс

asp2403 committed:
- В testingResult добавлено поле aborted. За счет этого убрано протаскивание параметра isTerminated в AppRunner и TestingServiceImpl

- Функционал, связанный с регистрацией пользователя вынесен в отдельный сервис UserService
- Добавлен интерфейс IOProvider и реализация IOProviderConsole, IOServiceImpl теперь принимает в конструкторе IOProvider

stvort commented on hw2/src/main/java/ru/otus/homework/popov/service/AppRunner.java:
```diff
@@ -2,120 +2,63 @@
-            if (s.equals(CMD_EXIT)) {
-                result = false;
-                break;
+            var testingResult = testingService.testUser(user);
```
Видимо что-то пошло не так при коммите. Тут передается только юзер

stvort commented on hw2/src/main/java/ru/otus/homework/popov/service/TestingService.java:
```diff
@@ -1,20 +1,14 @@
+public interface TestingService {
 
-    int getQuestionCount();
+    TestingResult testUser(User user, AtomicBoolean isTerminated);
```
А тут еще есть `AtomicBoolean `

stvort commented on hw2/src/main/java/ru/otus/homework/popov/service/TestingServiceImpl.java:
```diff
@@ -1,61 +1,55 @@
-        var answer = question.getAnswers().get(answerIndex);
-        if (answer.isCorrect()) {
-            user.addScore();
+    public TestingResult testUser(User user) {
```
А тут нет

asp2403 committed:
TestingService

stvort commented on hw2/src/main/java/ru/otus/homework/popov/service/AppRunner.java:
```diff
@@ -2,120 +2,63 @@
     private final IOService ioService;
-    private final QuestionConverter questionConverter;
-
+    private final AppConfig appConfig;
```
Это не используется

stvort commented on hw2/src/main/java/ru/otus/homework/popov/service/AppRunner.java:
```diff
@@ -2,120 +2,63 @@
-        run();
+        sayHello();
+        var user = userService.register();
+        user.ifPresent(this::run);
```
Лучше указать, что именно тут `run`

stvort commented on hw2/src/main/java/ru/otus/homework/popov/service/AppRunner.java:
```diff
@@ -2,120 +2,63 @@
+        this.userService = userService;
     }
 
     public void execute() {
```
А тут что именно `execute`

stvort commented on hw2/src/main/java/ru/otus/homework/popov/domain/Question.java:
```diff
@@ -4,11 +4,26 @@
 import java.util.Objects;
 
 public class Question {
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Question question = (Question) o;
+        return index == question.index && Objects.equals(body, question.body) && Objects.equals(answers, question.answers);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(index, body, answers);
+    }
```
Лучше придерживаться следующего порядка: сначала поля, потом конструктор, потом публичные методы, потом приватные

asp2403 committed:
Рефакторинг в соответствии с замечаниями
